pmlc/dialect/pxa/transforms/reorder_layout.cc

```diff
@@ -0,0 +1,264 @@
+// Copyright 2020 Intel Corporation
+
+#include <numeric>
+#include <vector>
+
+#include "mlir/Dialect/Affine/IR/AffineMemoryOpInterfaces.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/IR/IntegerSet.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/DebugStringHelper.h"
+
+#include "pmlc/dialect/pxa/transforms/pass_detail.h"
+#include "pmlc/dialect/pxa/transforms/passes.h"
+#include "pmlc/util/logging.h"
+
+namespace pmlc::dialect::pxa {
+
+namespace {
+
+class ReorderLayoutPass final : public ReorderLayoutBase<ReorderLayoutPass> {
+public:
+  void runOnFunction();
+};
+
+struct ReorderDesc {
+  mlir::AffineMap reorderMap;
+  mlir::SmallVector<uint64_t, 6> reorderedShape;
+  mlir::SmallVector<uint64_t, 6> reorderedVector;
+};
+
+/// Simplify affine map given integral constraints. Returns the same map if it
+/// cannot be simplified further.
+///
+/// Input:
+///   map         = (d0, d1) -> (d0 + d1, (d0 * 16 + d1 * 8) floordiv 8 floordiv
+///   2,
+///                              (d0 * 16 + d1 * 8) floordiv 8 % 2, 0)
+///   constraints = {0 <= d1 < 2, 0 <= d0 < 6}
+///
+/// Output:
+///   (d0, d1) -> (d0 + d1, d0, d1, 0)
+mlir::AffineMap simplifyConstrainedAffineMap(mlir::AffineMap map,
+                                             mlir::IntegerSet constraints) {
+  // TODO: Implement.
+  return map;
+}
+
+/// Expand affine map dimensions based on integral constraints and vector shape.
+///
+/// Input:
+///   map (A)     = (d0, d1) -> (d0 + d1, d0 * 16 + d1 * 8)
+///   shape       = <7, 96>
+///   vector      = <1, 8>
+///   constraints = {0 <= d1 < 2, 0 <= d0 < 6}
+/// Output:
+///   reorder map (B)  = (d0, d1) -> (d0, d1 floordiv 8 floordiv 2, d1 floordiv
+///   8 % 2, 0) reordered shape  = <7, 6, 2, 8> reordered vector = <1, 1, 1, 8>
+///
+/// Note: to obtain affine map from input space to expanded space composition
+///       A o B can be used (with simplification).
+/// A o B = (d0, d1) -> (d0 + d1, d0, d1, 0)
+ReorderDesc expandAffineMap(mlir::AffineMap map, mlir::ArrayRef<uint64_t> shape,
+                            mlir::ArrayRef<uint64_t> vector,
+                            mlir::IntegerSet constraints) {
+  // TODO: Implement.
+  auto reorderMap = mlir::AffineMap::getMultiDimIdentityMap(map.getNumResults(),
```

Comment:
I can see that what the example is doing, in that it is converting a the second dimension into components, and providing a mapping and new size/vector data.  However I don't understand what the overall goal in general is.  Is it to split all dimensions into the maximal number of 'independently strideable' bits or something like that?  I mean it's clear the the first dimension can't be 'tiled' in any useful way, and that the second dimension can be tiled into 3 parts, so maybe that makes sense?  But more description would be helpful.  Also, what would the algorithm for this be?

Reply:
I have added some additional description. General aim is to transform map of read operation to reduce number of input dimensions used for each output, ie (d0 * 16 + d1) uses two input dimensions, but in (d0, d1) both use one.
This is done in order to provide stronger hint for sorting where each dimension should be placed.

As for implementation again I  was thinking about rewrites with pattern matching, only one I see really needed after subgroups/tiling being e*i* * N + e*j*, where e*j* < N -> e*i*, e*j*